lib/prime.rb

```diff
@@ -32,15 +32,48 @@ def prime_division(generator = Prime::Generator23.new)
 
   # Returns true if +self+ is a prime number, else returns false.
   def prime?
-    return self >= 2 if self <= 3
-    return true if self == 5
-    return false unless 30.gcd(self) == 1
-    (7..Integer.sqrt(self)).step(30) do |p|
-      return false if
-        self%(p)    == 0 || self%(p+4)  == 0 || self%(p+6)  == 0 || self%(p+10) == 0 ||
-        self%(p+12) == 0 || self%(p+16) == 0 || self%(p+22) == 0 || self%(p+24) == 0
-    end
-    true
+    # https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test
+    # https://arxiv.org/pdf/1509.00864.pdf
+    #
+    # if n < 3,317,044,064,679,887,385,961,981,
+    # it is enough to test a = 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, and 41.
+    # and the result will be deterministic.
+    #
+
+    if self > 3317044064679887385961981
+      raise ArgumentError, ".prime? is only valid for values less than 3,317,044,064,679,887,385,961,981"
+    end
+    
+    return false if self < 2
+    return true  if self < 4
+    return false if self.even?
+
+    r = 0
+    d = self - 1
+
+    while d.even?
+      d /= 2
+      r += 1
+    end
+
+    [2,3,5,7,11,13,17,19,23,29,31,37,41].each do |a|
+      x = a.pow(d, self)
+      next if x == 1 || x == (self-1) || a == self
+
+      (r-1).times do
+        x = x.pow(2, self)
+        if x == (self-1)
+          break
+        end
+      end
+
+      if (x == self-1)
+        next
+      else
+        return false
+      end
```

Comment:
Okay, let's see what you think about this:
- I guess you are retesting ``x == self-1`` to avoid using a flag to break the outer ``each`` loop.
- However, ``break``ing ``times`` returns ``nil`` instead of the actual number, ``r-1`` in this case.
- So you could use the return value of ``times`` to simplify this code block to the following
```Ruby
      return false if r.times do # I replaced (r-1) with r, see my previous comment
        x = x.pow(2, self)
        break if x == (self-1)
      end
```

It is less readable because you need to know about the return value of a breaking block, but it works at least from 0 to 25000 (I can't test a bigger set right now, sorry).
The idea here is to avoid repeating ``(x == self-1)``, feel free to ignore this idea if you think it would impact readability, you can use the simpler ``return false if x != self-1`` after the ``times`` loop and make ``next`` implicit.

Reply:
I don't think Miller-Rabin will be changing that much, so I swapped this in.